pkg/synccompactor/compactor.go

```diff
@@ -0,0 +1,124 @@
+package synccompactor
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+
+	reader_v2 "github.com/conductorone/baton-sdk/pb/c1/reader/v2"
+	"github.com/conductorone/baton-sdk/pkg/dotc1z"
+	c1zmanager "github.com/conductorone/baton-sdk/pkg/dotc1z/manager"
+	sync_compactor "github.com/conductorone/baton-sdk/pkg/synccompactor/naive"
+)
+
+type Compactor struct {
+	entries []*CompactableSync
+	destDir string
+}
+
+type CompactableSync struct {
+	filePath string
+	syncID   string
+}
+
+func NewCompactor(ctx context.Context, destDir string, compactableSyncs ...*CompactableSync) (*Compactor, error) {
+	if len(compactableSyncs) == 0 || len(compactableSyncs) == 1 {
+		return nil, errors.New("must provide two or more files to Compact")
+	}
+
+	return &Compactor{entries: compactableSyncs, destDir: destDir}, nil
+}
+
+func (c *Compactor) Compact(ctx context.Context) (*CompactableSync, error) {
+	if len(c.entries) < 2 {
+		return nil, nil
+	}
+
+	tempDir := os.TempDir()
+
+	base := c.entries[0]
+	var latest *CompactableSync
+	for i := 1; i < len(c.entries); i++ {
+		applied := c.entries[i]
+
+		compactable, err := c.doOneCompaction(ctx, tempDir, base, applied)
+		if err != nil {
+			return nil, err
+		}
+		latest = compactable
+		base = compactable
+	}
+
+	return latest, nil
+}
+
+func getLatestObjects(ctx context.Context, info *CompactableSync) (*reader_v2.SyncRun, *dotc1z.C1File, c1zmanager.Manager, func(), error) {
+	baseC1Z, err := c1zmanager.New(ctx, info.filePath)
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	baseFile, err := baseC1Z.LoadC1Z(ctx)
+	if err != nil {
+		return nil, nil, nil, nil, err
+	}
+
+	latestAppliedSync, err := baseFile.GetSync(ctx, &reader_v2.SyncsReaderServiceGetSyncRequest{
+		SyncId:      info.syncID,
+		Annotations: nil,
+	})
```

Comment:
If baseFile.GetSync() errors (if the sync ID doesn't exist, for example), we don't close baseFile or baseC1Z.

Reply:
Fixed